Migrate user card popovers to tippy.

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
This PR is still a Work in progress. The PR description will be updated soon.

Fixes: <!-- Issue link, or clear description.-->
#24280
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Right sidebar user card</summary>
<br>

![image](https://user-images.githubusercontent.com/64723994/236819420-ec5a2e79-c3f4-411e-9847-268e92bdba86.png)

![image](https://user-images.githubusercontent.com/64723994/236812840-ebaedc74-98e7-43c0-aa75-2e60cf3fa876.png)

</details>

<details>
<summary>Recent topic user card</summary>
<br>

![image](https://user-images.githubusercontent.com/64723994/236813062-71f0852a-35db-4989-b036-b76fcbb9869f.png)

With fallback placement left when there is not enough space on right:
![image](https://user-images.githubusercontent.com/64723994/236813185-395d7830-3a64-4c64-9c0a-7d2f26ba91a9.png)

</details>

<details>
<summary>.veiw_user_profile user card</summary>
<br>

![image](https://user-images.githubusercontent.com/64723994/236813690-cca349ee-3d97-46fb-ab60-3734f0ff97fe.png)

</details>

<details>
<summary>Message row related user cards</summary>
<br>

- From Sender Name: 
 ![image](https://user-images.githubusercontent.com/64723994/236814314-e7f5b2b9-21e5-4f5b-921c-c8d5206c28b3.png)

- From user avatar: 
![image](https://user-images.githubusercontent.com/64723994/236814601-c712fad2-f085-4382-a506-a77a371ce93d.png)

- From `U` keyboard shortcut:
![image](https://user-images.githubusercontent.com/64723994/236814829-63b43412-7c52-4bcc-84c0-c38f9118675a.png)
  - with fallback placement:
  ![image](https://user-images.githubusercontent.com/64723994/236814942-6d4fb005-1d9f-4fe4-baab-40ff18c87751.png)

- user mentions pill:
  ![image](https://user-images.githubusercontent.com/64723994/236815200-73e79a36-43da-41fb-92e4-8b6d2667c35f.png)

</details>


<details>
<summary>On device width less than 576px, I kept all user cards the same way, centered with background overlays for consistency (previously right sidebar user cards had different behavior) [Feedback required for this change]</summary>
<br>

![image](https://user-images.githubusercontent.com/64723994/236816138-bbbf6e7e-e42f-46ba-8c3e-b7b0738a91ce.png)

![image](https://user-images.githubusercontent.com/64723994/236816206-70c7a79c-cb4c-4e16-a1f5-e11b2d31499b.png)

![image](https://user-images.githubusercontent.com/64723994/236816273-6243d44d-0692-460f-9005-65f15ed2c622.png)

</details>


Video showing the whole behavior:

https://user-images.githubusercontent.com/64723994/236818630-28a11dcf-03bb-40ab-bf59-c080390e8dcc.mp4


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ddharmani3 commented:
**Help Wanted:**

1. The user card does not open with a single click on the user mention pills; instead, it requires a double click.

2. User cards opened through message view (i.e., by clicking on the sender name, user avatar, or user mention) do not open consistently. There are two ways to open them that I am aware of:

  1. Double-click on the user mention, which will open the user card. Then, when the user card is open, clicking on the sender name or user avatar will also open the user card.

  2. Double-tap on the message, which will open a user card. Then, when that user card is open, clicking on the sender name or user avatar will also open the user card.

  - Here is a video demonstrating both methods:
  
https://user-images.githubusercontent.com/64723994/236898341-b477d3cb-2a23-4d3b-9a0b-0c30e7ecb61d.mp4

Maybe the reason for this issue can be that click handlers are not properly attached to these elements as the only failing test is this:
```
test failed: /__w/zulip/zulip/web/tests/popovers.test.js > sender_hover

AssertionError [ERR_ASSERTION]: no click handler for #main_div .sender_name, .sender_name-in-status, .inline_profile_picture
```

3. In a Bot's user card the bot owner's name is clickable which opens the bot owner's user card, this is not working in my PR.

Any help on debugging any of the above bugs is much appreciated.

Ddharmani3 commented:
**Feedback wanted:**
1. Now all the user cards disappear on scroll, previously only message-info and recent conversations popovers had this behavior.
2. Now all the user cards on the device width < 576px pops in the center with a background overlay, previously only the right sidebar didn't have this behavior.
The above 2 points make user card behavior consistent in all user cards.

alya commented:
> **Feedback wanted:**
> 
> 1. Now all the user cards disappear on scroll, previously only message-info and recent conversations popovers had this behavior.
> 2. Now all the user cards on the device width < 576px pops in the center with a background overlay, previously only the right sidebar didn't have this behavior.
>    The above 2 points make user card behavior consistent in all user cards.

Both of these changes seem reasonable to me.

shameondev commented:
I believe the solution to the issues with clicks not working may lie in the element to which we attach event listeners. Previously, the listeners were attached to `#main_div`, but now `tippy_no_propagation` attaches everything to `body`.

Therefore, I think one possible solution could be to use the `delegate` function from `Tippy` and delegate it to `#main_div`.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -907,4 +920,265 @@ export function initialize() {
+    }
+
+    function placement_on_mobile(instance) {
+        if ($("body").outerWidth() < 576) {
+            const $overlay = $("<div>").addClass("tippy-overlay").css({
+                position: "fixed",
+                top: "0",
+                left: "0",
+                width: "100%",
+                height: "100%",
+                backgroundColor: "rgba(0, 0, 0, 0.7)",
+                zIndex: 9998,
+            });
+
+            $("body").append($overlay);
+            $overlay.on("click", () => {
+                $overlay.remove();
+            });
+            const $user_card_tippy = $(instance.popper);
+            const left = ($("body").innerWidth() - $user_card_tippy.innerWidth()) / 2;
+            const top = ($("body").outerHeight() - $user_card_tippy.outerHeight()) / 2;
+            $user_card_tippy.css({
+                transform: `translate(${left}px, ${top}px)`,
+            });
```
This doesn't seem to be working. (Maybe due to the recent changes in main)

I don't think this is a good way to do this though. Can you look into if there are methods you can use to get the same behaviour without using a separate `div` here?

amanagr commented:
![image](https://github.com/zulip/zulip/assets/25124304/6bb73670-323b-4da1-9dbd-e9f19942eddb)

There is extra space around the box which we don't need. You can use negative margin here to remove them since they were added by tippy.

shameondev commented:
Hey, @Ddharmani3! Thanks for your contributions! Would you mind summarizing the changes you made in the last iteration? It would greatly assist the reviewer in their review process.

Ddharmani3 commented:
> Hey, @Ddharmani3! Thanks for your contributions! Would you mind summarizing the changes you made in the last iteration? It would greatly assist the reviewer in their review process.

**Summary:**

- This PR converts user card popover to tippy this will be followed by a redesign of the user card suggested in #22802 in a separate PR.

- The issues mentioned [above](https://github.com/zulip/zulip/pull/25459#issuecomment-1538821102) is fixed in the updated PR.
 
- I am not sure why @amanagr got the above-mentioned issues I was not getting those issues locally. Can you check if you are getting those issues?

- There is one common object `user_card_options` that contains `onShow`, `onMount`, and `onHidden` functions that are used by all the tippy. The uncommon/unique options for each popover are set using `instance.context` inside the `onCreate` function for each tippy individually.

- `tippy_popover` class is common to all popovers. There is another class on popovers that is `user_popover` for all popovers except buddy list user info popovers which have class `sidebar_user_popover`.

This PR is ready for review @shameondev @amanagr @ryanreh99.

ryanreh99 commented on web/templates/user_info_popover_content.hbs:
```diff
@@ -1,172 +1,191 @@
+                        {{#if invisible_mode}}
+                        <li>
+                            <a tabindex="0" class="invisible_mode_turn_off">
+                                <i class="fa fa-circle-o" aria-hidden="true"></i> {{t "Turn off invisible mode" }}
+                            </a>
+                        </li>
+                        {{else}}
+                        <li>
+                            <a tabindex="0" class="invisible_mode_turn_on">
+                                <i class="fa fa-circle-o" aria-hidden="true"></i> {{t "Go invisible" }}
```
Since we're updating the invisible mode icons to `zulip-icon-play-circle` and `zulip-icon-play-circle` in profile menu.
we would want to update the icons here as well.

Ddharmani3 replied:
@ryanreh99, Here is the redesign PR if you want to take a look: #26170

Ddharmani3 commented:
Resolved conflicts there is one issue with the user card don't open on clicking the sender name in the ` /me message` due to its layout (I don't know the exact reason but removing `display: inline` from `.sender-status` it becomes `display: flex` due to another CSS and the click works) It is working fine on snder_name on normal message.

Waiting for undergoing [layout change](https://chat.zulip.org/#narrow/stream/6-frontend/topic/me-message.20layout) of `/me message` to see if it solves the issue automatically with the new layout. Meanwhile, I will also try to find the reason and solution for it, as I am curious too how the layout is affecting the click handler.

amanagr commented:
@Ddharmani3 sorry for missing the ping earlier. Can you please rebase it, I want to review this.

Ddharmani3 commented:
> @Ddharmani3 sorry for missing the ping earlier. Can you please rebase it, I want to review this.

sure, I will rebase it soon

Ddharmani3 commented:
@amanagr, I have rebased this PR. This PR is ready for review now.

amanagr commented on web/styles/popovers.css:
```diff
@@ -56,6 +56,75 @@
     }
 }
 
+.tippy_popover {
```
If you want to definite CSS specific to user card popover, then this is not a good classname.

amanagr commented on web/styles/popovers.css:
```diff
@@ -56,6 +56,75 @@
+    }
+}
+
+[data-tippy-root]:has(.tippy_popover) {
```
What are  we using `:has` here?

amanagr commented on web/src/popover_menus.js:
```diff
@@ -918,4 +931,268 @@ export function initialize() {
             popover_instances.send_later = undefined;
         },
     });
+
```
It is hard to verify the changes below since they are being deleted and added in separate commits. It would be better to just have them in one commit.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -1253,4 +1253,28 @@ export function initialize() {
         },
         ...user_card_options,
     });
+
+    const message_user_info_popovers = [".sender_name_user_card", ".profile_picture"];
+
+    for (const target of message_user_info_popovers) {
```
Maybe just past them both as `target`. A `for` loop here doesn't look good.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -155,6 +156,28 @@ function get_popover_items_for_instance(instance) {
     return $current_elem.find("li:not(.divider):visible a");
 }
 
+export function show_sender_info() {
```
Should we name this `toggle_user_card_popover_from_hotkey`?

amanagr commented:
@Ddharmani3 looks great! Left some comments.

Ddharmani3 commented:
> @Ddharmani3 looks great! Left some comments.

@amanagr addressed the feedback. This PR is ready for review now.
Didn't understand the [commit feedback](https://github.com/zulip/zulip/pull/25459#discussion_r1251682205) properly, but tried to implement it as per what I think you are suggesting.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -918,4 +931,258 @@ export function initialize() {
+            const user = people.get_by_user_id(user_id);
+            user_profile.show_user_profile(user);
+        });
+        popover_menus_data.init_email_clipboard();
+        popover_menus_data.init_email_tooltip(user);
```
Let's move this code above `popover_menus_data.clipboard_enable`.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -918,4 +931,258 @@ export function initialize() {
+            const content = $(render_user_info_popover_content(args)).get(0);
+            instance.setContent(parse_html(content.innerHTML));
+            instance.setProps({
+                popperOptions: {
+                    strategy: "fixed",
+                    modifiers: [
+                        {
+                            name: "preventOverflow",
+                            options: {
+                                altAxis: true,
+                                padding: {top: 10, bottom: $("#compose").outerHeight()},
+                            },
+                        },
+                        {
+                            name: "eventListeners",
+                            options: {
+                                scroll: false,
+                                resize: false,
+                            },
```
Can you add some comments here explaning your choices.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -918,4 +931,258 @@ export function initialize() {
+            $(".simplebar-content-wrapper").off("scroll");
+            instance.destroy();
+            popover_instances.user_card = undefined;
+            $(".tippy-overlay").remove();
```
`$(".tippy-overlay")` that's a very generic class name to remove with `onHidden`. We definitely should localize the name to represent user card popover overlay.

amanagr commented:
@Ddharmani3 left some more comments. Thanks for the update!

Ddharmani3 commented:
@amanagr, addressed all the feedback, this PR is ready for review now.

amanagr commented on web/src/popover_menus.js:
```diff
@@ -918,4 +931,258 @@ export function initialize() {
+        popover_menus_data.clipboard_enable($(".copy_mention_syntax")[0]);
+
+        $popper.one("click", ".copy_mention_syntax", () => {
+            popovers.hide_all();
```
Oh! I mis-remembered what `hideOnclick` does but still we should be doing better here.

amanagr commented:
![user card popover](https://github.com/zulip/zulip/assets/25124304/bafcf8a4-43f4-4737-8316-10449b390e99)
The "unmute/managae user popover" seems to bugged and doesn't close on click.

Ddharmani3 commented:
@amanagr the bug was due to removing popvers.hideALL(), added it back where needed. This PR is ready for review now.

amanagr commented:
@Ddharmani3 sorry I missed your last ping. Can you rebase this?

shameondev commented:
Hey, @Ddharmani3, would you mind if I step in and make the necessary changes to your work here?

Ddharmani3 commented:
> Hey, @Ddharmani3, would you mind if I step in and make the necessary changes to your work here? This is an important project that needs to be finished soon.

@shameondev , sure if you can do that it would be a great help as I already have some PRs in the pipeline to work on before this, and with my reduced working hour it would take me 1-2 weeks to get this ready for integration review.

If you need any help understanding the code or if you want to collaborate on some parts of PR, 
DM me on CZO, I will be more than happy to help :)

timabbott commented:
Closing as completed through another part of https://github.com/zulip/zulip/issues/23632, specifically #25459.
